File: frontend/src/services/realtime.ts
export const realtimeEventTypes = ['SYSTEM_STATUS', 'INGEST_METRICS', 'REPLAY_STATUS', 'FEATURE_METRICS', 'DETECTION_RESULT', 'ALERT_CREATED', 'ALERT_UPDATED'] as const;
export type RealtimeEventType = typeof realtimeEventTypes[number];

export function isRealtimeEventType(value: unknown): value is RealtimeEventType {
  return typeof value === 'string' && (realtimeEventTypes as readonly string[]).includes(value);
}

File: backend/app/schemas/realtime.py
"""Realtime event envelope for browser consumers."""

from datetime import datetime
from typing import Any, Literal
from pydantic import BaseModel


class RealtimeEnvelope(BaseModel):
    event_type: Literal[
        "SYSTEM_STATUS", "INGEST_METRICS", "REPLAY_STATUS", "FEATURE_METRICS",
        "heartbeat", "alert", "system_status", "DETECTION_RESULT", "ALERT_CREATED", "ALERT_UPDATED",
    ]
    timestamp: datetime
    data: Any
